google-cloud-pubsub/lib/google/cloud/pubsub/async_publisher/batch.rb

```diff
@@ -0,0 +1,293 @@
+          # @return [Symbol] The state of the batch.
+          #
+          #   * `:added` - Added to the active list of items to be published.
+          #   * `:queued` - Batch is publishing, and the messsage is queued.
+          #   * `:full` - Batch is full and ready to be published, and the
+          #     message is queued.
+          #
+          def add msg, callback
+            synchronize do
+              raise AsyncPublisherStopped if @stopping
+              raise OrderingKeyCanceled, @ordering_key if @canceled
+
+              if @publishing
+                queue_add msg, callback
+                :queued
+              elsif try_add msg, callback
+                :added
+              else
+                queue_add msg, callback
+                :full
+              end
+            end
+          end
+
+          ##
+          # Marks the batch to be published.
+          #
+          # The method will indicate whether a new publishing job should be
+          # started to publish the batch. See {publishing?}.
+          #
+          # @param [Boolean] stop Indicates whether the batch should also be
+          #   marked for stopping, and any existing publish job should publish
+          #   all items until the batch is empty.
+          #
+          # @return [Boolean] Returns whether a new publishing job should be
+          #   started to publish the batch. If the batch is already being
+          #   published then this will return `false`.
+          #
+          def publish! stop: nil
+            synchronize do
+              @stopping = true if stop
+
+              return false if @canceled
+
+              # If we are already publishing, do not indicate a new job needs to
+              # be started.
+              return false if @publishing
+
+              @publishing = !(@items.empty? && @queue.empty?)
+            end
+          end
+
+          ##
+          # Indicates whether the batch has an active publishing job.
+          #
+          # @return [Boolean]
+          #
+          def publishing?
+            # This probably does not need to be synchronized
+            @publishing
+          end
+
+          ##
+          # Indicates whether the batch has been stopped and all items will be
+          # published until the batch is empty.
+          #
+          # @return [Boolean]
+          #
+          def stopping?
+            # This does not need to be synchronized because nothing un-stops
+            @stopping
+          end
+
+          ##
+          # Rebalances the batch by moving any queued items that will fit into
+          # the active item list.
+          #
+          # This method is only intended to be used by the active publishing
+          # job.
+          #
+          def rebalance!
+            synchronize do
+              return [] if @canceled
+
+              until @queue.empty?
+                item = @queue.first
+                if try_add item.msg, item.callback
```

Comment:
What happens if I call publish on a message that is larger than the max batch size I have set? It looks like publishing will just just stop because we'll never be able to add it. In other languages, we consider it okay to publish this message in a batch with just that message.

Reply:
Yes, this is a regression. I'll update the implementation to restore the logic that allowed for this.

Reply:
Fixed.